query/analyze/structure.feature

```diff
@@ -0,0 +1,300 @@
+      """
+      match
+        entity $p1;
+        $p1 sub! person, owns $n1;
+        $n1 label name;
+        $n1 value string;
+        $f sub friendship, relates friend;
+        $p1 plays friendship:friend;
+        $p2 plays $role;
+      """
+    Then analyzed query pipeline structure is:
+    """
+    Pipeline([
+      Match([
+        Kind(entity, $p1),
+        SubExact($p1, person), Owns($p1, $n1),
+        Label($n1, name),
+        Value($n1, string),
+        Sub($f, friendship), Relates($f, friend),
+        Plays($p1, friendship:friend),
+        Plays($p2, $role)
+      ])
+    ])
+    """
+    Given transaction closes
+
+
+  Scenario: All stages in a pipeline are present in the structure
+    Given connection open read transaction for database: typedb
+    When get answers of typeql analyze query
+    """
+    match
+     $p isa person;
+     $q isa person;
+     $n isa name == "John";
+    insert
+     $p has $n;
+    delete
+      has $n of $p;
+      $q;
+    update
+      $p has $n;
+    put
+      $p has $n;
+    distinct;
+    match
+      try { $p has ref $ref; };
+    require $ref;
+    select $ref, $n;
+    reduce $ref_sum = sum($ref) groupby $n;
+    sort $n desc;
+    offset 1;
+    limit 1;
+    """
+    Then analyzed query pipeline structure is:
+    """
+    Pipeline([
+      Match(
+        [Isa($p, person), Isa($q, person), Isa($n, name), Comparison($n, "John", ==)]
+      ),
+      Insert([Has($p, $n)]),
+      Delete([$q], [Has($p, $n)]),
+      Update([Has($p, $n)]),
+      Put([Has($p, $n)]),
+      Distinct(),
+      Match([
+        Try([Has($p, $ref), Isa($ref, ref)])
+      ]),
+      Require([$ref]),
+      Select([$n, $ref]),
+      Reduce(
+        [ReduceAssign($ref_sum, Reducer(sum, [$ref]))],
+        [$n]
+      ),
+      Sort([desc($n)]),
+      Offset(1),
+      Limit(1)
+    ])
+    """
+    Given transaction closes
+
+  # Unhappy path
+
+  Scenario: Unsatisfiable schema queries still return
+    Given connection open schema transaction for database: typedb
+    Given typeql schema query
+      """
+      define
+      relation new-relation, relates new-role;
+      """
+    Given transaction commits
+
+    Given connection open read transaction for database: typedb
+    When get answers of typeql analyze query
+    """
+    match
+     $p sub! person;
+    """
+    Then analyzed query pipeline structure is:
+    """
+    Pipeline([
+      Match(
+        [SubExact($p, person)]
+      )
+    ])
+    """
+
+    When get answers of typeql analyze query
+    """
+    match
+     $r sub friendship, relates new-role;
+    """
+    Then analyzed query pipeline structure is:
+    """
+    Pipeline([
+      Match(
+        [Sub($r, friendship), Relates($r, new-role)]
+      )
+    ])
+    """
+
+    Given transaction closes
+
+
+  Scenario: Errors in the query are returned as errors
+    Given connection open read transaction for database: typedb
+    When typeql analyze query; parsing fails
```

Comment:
I think Cole's commentary is reflecting something accurate: `analyze()` shouldn't throw an error, but it should return an error. I think we do unify this in the service layer anyway right? So this could be generically "fails with message containing..."?